Sources/SotoCore/Credential/ConfigFileCredentialProvider.swift

```diff
@@ -43,82 +31,91 @@ class AWSConfigFileCredentialProvider: CredentialProviderSelector {
+        let profile = profile ?? Environment["AWS_PROFILE"] ?? ConfigFileLoader.default
         let threadPool = NIOThreadPool(numberOfThreads: 1)
         threadPool.start()
         let fileIO = NonBlockingFileIO(threadPool: threadPool)
 
-        Self.getSharedCredentialsFromDisk(credentialsFilePath: credentialsFilePath, profile: profile, on: eventLoop, using: fileIO)
+        Self.getSharedCredentialsFromDisk(credentialsFilePath: credentialsFilePath, configFilePath: configFilePath,
+                                          profile: profile, context: context, using: fileIO)
             .always { _ in
                 // shutdown the threadpool async
                 threadPool.shutdownGracefully { _ in }
             }
             .cascade(to: self.startupPromise)
     }
 
+    /// Load credentials from disk
+    /// - Parameters:
+    ///   - credentialsFilePath: file path for AWS credentials file
+    ///   - configFilePath: file path for AWS config file (optional)
+    ///   - profile: named profile to load (optional)
+    ///   - context: credential provider factory context
+    ///   - fileIO: non-blocking file IO
+    /// - Returns: Promise of a Credential Provider (StaticCredentials or STSAssumeRole)
     static func getSharedCredentialsFromDisk(
         credentialsFilePath: String,
+        configFilePath: String?,
         profile: String,
-        on eventLoop: EventLoop,
+        context: CredentialProviderFactory.Context,
         using fileIO: NonBlockingFileIO
     ) -> EventLoopFuture<CredentialProvider> {
-        let filePath = Self.expandTildeInFilePath(credentialsFilePath)
 
-        return fileIO.openFile(path: filePath, eventLoop: eventLoop)
+        return loadFile(path: credentialsFilePath, on: context.eventLoop, using: fileIO)
+            .flatMap { credentialsByteBuffer -> EventLoopFuture<(ByteBuffer, ByteBuffer?)> in
+                if let path = configFilePath {
+                    return Self.loadFile(path: path, on: context.eventLoop, using: fileIO).map { (credentialsByteBuffer, $0) }
+                }
+                else {
+                    return context.eventLoop.makeSucceededFuture((credentialsByteBuffer, nil))
+                }
+            }
+            .flatMapThrowing { credentialsByteBuffer, configByteBuffer in
+                return try ConfigFileLoader.sharedCredentials(from: credentialsByteBuffer, configByteBuffer: configByteBuffer,
+                                                              for: profile, context: context)
+            }
+    }
+
+    /// Load a file from disk without blocking the current thread
+    /// - Parameters:
+    ///   - path: path for the file to load
+    ///   - eventLoop: event loop to run everything on
+    ///   - fileIO: non-blocking file IO
+    /// - Returns: Event loop future with file contents in a byte-buffer
+    static func loadFile(path: String, on eventLoop: EventLoop, using fileIO: NonBlockingFileIO) -> EventLoopFuture<ByteBuffer> {
```

Comment:
You had `loadFile` in ConfigFileLoader.swift for a while any reason you returned it to ConfigFileCredentialProvider.swift

Reply:
I agree with the ownership model. I'll move all `fileIO` related code to `ConfigFileLoader`.